File: sdk/android/EmberClient.kt
package ember

import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

class EmberClient(
    private val projectId: String,
    private val apiKey: String,
    private val ingestUrl: String,
) {
    init {
        require(projectId.isNotBlank()) { "projectId requis" }
        require(apiKey.isNotBlank()) { "apiKey requis" }
        require(ingestUrl.isNotBlank()) { "ingestUrl requis" }
    }

    fun captureMessage(level: String, message: String, tags: Map<String, String>? = null, release: String? = null, env: String? = null) {
        if (message.isBlank()) return
        val payload = buildPayload(level.ifBlank { "info" }, "Message", message, tags, release, env)
        send(payload)
    }

    fun captureError(error: Throwable, tags: Map<String, String>? = null, release: String? = null, env: String? = null) {
        val message = error.message ?: "Erreur inconnue"
        val payload = buildPayload("error", error.javaClass.simpleName, message, tags, release, env)
        send(payload)
    }

    private fun buildPayload(level: String, exceptionType: String, message: String, tags: Map<String, String>?, release: String?, env: String?): String {
        val context = mutableListOf<String>()
        tags?.takeIf { it.isNotEmpty() }?.let {
            val tagsJson = it.entries.joinToString(",") { entry -> "\"${entry.key}\":\"${entry.value}\"" }
            context.add("\"tags\":{${tagsJson}}")
        }
        release?.takeIf { it.isNotBlank() }?.let { context.add("\"release\":\"$it\"") }
        env?.takeIf { it.isNotBlank() }?.let { context.add("\"env\":\"$it\"") }

        val contextJson = if (context.isNotEmpty()) ",\"context\":{${context.joinToString(",")}}" else ""

        return "{" +
            "\"event_id\":\"${UUID.randomUUID()}\"," +
            "\"project_id\":\"$projectId\"," +
            "\"timestamp\":\"${java.time.Instant.now()}\"," +
            "\"level\":\"$level\"," +
            "\"message\":\"${escape(message)}\"," +
            "\"exception\":{\"type\":\"${escape(exceptionType)}\",\"message\":\"${escape(message)}\"}," +
            "\"sdk\":{\"name\":\"ember-android\",\"version\":\"0.1.0\"}" +
            contextJson +
            "}"
    }

    private fun send(payload: String) {
        val url = URL(ingestUrl.trimEnd('/') + "/ingest")
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.setRequestProperty("x-ember-key", apiKey)
        connection.doOutput = true
        OutputStreamWriter(connection.outputStream).use { writer ->
            writer.write(payload)
        }
        connection.inputStream.close()
        connection.disconnect()
    }

    private fun escape(value: String): String {
        return value.replace("\\", "\\\\").replace("\"", "\\\"")
    }
}

File: sdk/ios/EmberClient.swift
import Foundation

public final class EmberClient {
    private let projectId: String
    private let apiKey: String
    private let ingestUrl: String

    public init(projectId: String, apiKey: String, ingestUrl: String) {
        self.projectId = projectId
        self.apiKey = apiKey
        self.ingestUrl = ingestUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    public func captureMessage(level: String, message: String, tags: [String: String]? = nil, release: String? = nil, env: String? = nil) {
        guard !message.isEmpty else { return }
        let payload = buildPayload(level: level.isEmpty ? "info" : level, exceptionType: "Message", message: message, tags: tags, release: release, env: env)
        send(payload: payload)
    }

    public func captureError(_ error: Error, tags: [String: String]? = nil, release: String? = nil, env: String? = nil) {
        let message = (error as NSError).localizedDescription
        let payload = buildPayload(level: "error", exceptionType: String(describing: type(of: error)), message: message, tags: tags, release: release, env: env)
        send(payload: payload)
    }

    private func buildPayload(level: String, exceptionType: String, message: String, tags: [String: String]?, release: String?, env: String?) -> [String: Any] {
        var payload: [String: Any] = [
            "event_id": UUID().uuidString,
            "project_id": projectId,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
            "level": level,
            "message": message,
            "exception": [
                "type": exceptionType,
                "message": message
            ],
            "sdk": [
                "name": "ember-ios",
                "version": "0.1.0"
            ]
        ]

        var context: [String: Any] = [:]
        if let tags = tags, !tags.isEmpty { context["tags"] = tags }
        if let release = release, !release.isEmpty { context["release"] = release }
        if let env = env, !env.isEmpty { context["env"] = env }
        if !context.isEmpty { payload["context"] = context }

        return payload
    }

    private func send(payload: [String: Any]) {
        guard let url = URL(string: ingestUrl + "/ingest") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-ember-key")

        let data = try? JSONSerialization.data(withJSONObject: payload)
        request.httpBody = data

        URLSession.shared.dataTask(with: request).resume()
    }
}
